Add SoftMaxNode/Symbol

fastbodin opened the pull request:
Fix #281. Basic implementation of softmax (https://en.wikipedia.org/wiki/Softmax_function) without axis arguments. 

I am concerned about errors in doubles due to rounding over time but thought it would be best to get feedback on implementation as is.

wbernoudy commented on dwave/optimization/mathematical.py:
```diff
@@ -1237,6 +1239,39 @@ def safe_divide(x1: ArraySymbol, x2: ArraySymbol) -> SafeDivide:
+        >>> with model.lock():
+        ...     model.states.resize(1)
+        ...     i.set_state(0, [1, 3, 2])
+        ...     print(sum(np.isclose(sm.state(0), expected)))
```
Minor nit but I might frame this as `print(np.isclose(sm.state(), expected).all())`

wbernoudy commented on tests/cpp/nodes/mathematical/test_softmax.cpp:
```diff
@@ -0,0 +1,170 @@
+            THEN("The initial softmax state is correct") {
+                // softmax should be ~ [0.00216569646006 0.00588697333334
+                // 0.11824302025266 0.87370430995393]
+                CHECK(softmax_ptr->view(state)[0] > 0.00216569646);
```
I think you'll want to use `CHECK_THAT(..., WithinRel(...));` for these

fastbodin replied:
Much better that what I was doing. Thanks!

fastbodin commented:
Made above suggested changes, fixed a few typos, and added a test for when resizing occurs.

wbernoudy commented on dwave/optimization/include/dwave-optimization/nodes/mathematical.hpp:
```diff
@@ -364,6 +364,60 @@ using MinNode = ReduceNode<functional::min<double>>;
+    /// @copydoc Array::sizeinfo()
+    SizeInfo sizeinfo() const override;
+
+    double get_denominator(const State& state) const;
```
I think this method should be private

wbernoudy replied:
IMO there's no need to test the denominator directly in the unit tests, I would just remove those

fastbodin replied:
Removed. To make up for the fact that we are no longer testing it directly, I added a test which 
- `commits` a `state`,
- `propagates` a `state`,
- `reverts` a  `state`, and
- `propagates` a `state`.

This can only result in the correct values of SoftMaxNode at the end if the denominator is properly reset in revert.

wbernoudy commented on dwave/optimization/mathematical.py:
```diff
@@ -1237,6 +1239,39 @@ def safe_divide(x1: ArraySymbol, x2: ArraySymbol) -> SafeDivide:
+    See Also:
+        :class:`~dwave.optimization.symbols.SoftMax`: equivalent symbol.
+
+    .. versionadded:: 0.6.4
```
We'll have to bump these to 0.6.5 as 0.6.4 was just released, sorry!

wbernoudy commented on dwave/optimization/src/nodes/mathematical.cpp:
```diff
@@ -1671,6 +1670,147 @@ template class ReduceNode<std::logical_or<double>>;
+            new_denominator += exp_val - node_data->get(u.index) * prior_denominator;
+            // Temporarily store in format of unmodified elements. Will multiply
+            // all elements by prior_denominator / new_denominator after.
+            node_data->set(u.index, exp_val / prior_denominator, true);
```
One issue I see with using `set()` here is that it will cause an `Update` to be emitted on each changed/placed value of the predecessor, but then if the denominator has changed (which it very likely has if there are any updates to the pred array), another `Update` will be emitted when the value is adjusted for the new denominator.

I'm thinking that in this loop, you should not use `set()` to avoid emitting these extraneous updates, and then use `set()` as you update them for the new denominator. This would mean you can't (easily) do the check to see if the new denominator is different from the prior denominator and skip emitting the updates if so.

fastbodin replied:
I forgot that doing this could result in two updates for the same index. This makes extra work for any predecessor. Good point.

fastbodin replied:
I used `std::vector` to track the indices I need to divide by the new denominator and to record the `exp()` values. I could alternatively use `std::unordered_map`. Recommendations for which is best for this scenario?

fastbodin commented:
Corrections include:
1. Removed SoftMaxNode::get_denominator and relevant C++ tests. Added a C++ test to compensate.
2.  Opted to not implement a method to check if `prior_denominator == new_denominator` (which could reduce the # of updates) since this edge case is unlikely in practical use cases. 
3. Opted to continue to cache updated `exp()` in `std::unordered_map` (as opposed to recording the indices that have been updated and recompute the value at these indices fully with an additional `exp()` call) because there was a bug that occurred when an update was called more than once on the same index. This choice increases memory usage but should result in faster computations.
